Add the integration of example-solution tests in ci

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'ci'

**Context**: 

See #1856 and #1850 

**Summary**:

Be able to:
 - Spawn two nodes (Bootstrap node with MetalK8s installed and a second node1)
 - Prepare node1 with Cypress deps
 - Add example-solution iso file built from this PR (https://github.com/scality/metalk8s/pull/2312)
 - Deploy Solution Operator in some Namespace (example-solution)
 - Apply example-solution CRs


**Acceptance criteria**: 

Debts on : https://github.com/scality/metalk8s/pull/2315 since the Pod containers cannot startup because of missing binaries 
and https://github.com/scality/metalk8s/pull/2320 since we require a Kubeconfig to run any kubectl command.

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1850

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1856-add-Integration-of-Solution-tests-in-CI` into
`development/2.4`
* [w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI) will be merged into `development/2.5` 
* [w/2.6/feature/1856-add-Integration-of-Solution-tests-in-CI](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/1856-add-Integration-of-Solution-tests-in-CI) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #2bb85d586f9022b6e8e0d9ea42b5d00668293365 on the integration branch
`w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI` is merging a branch which is neither the current
branch `feature/1856-add-Integration-of-Solution-tests-in-CI` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/1856-add-Integration-of-Solution-tests-in-CI` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #90b9cd24910478f9d54192ca0df45ab9e7bc73c8 on the integration branch
`w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI` is merging a branch which is neither the current
branch `feature/1856-add-Integration-of-Solution-tests-in-CI` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/1856-add-Integration-of-Solution-tests-in-CI` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented on eve/main.yml:
```diff
@@ -1466,3 +1469,277 @@ stages:
+          command: >
+            ssh -F ssh_config bootstrap
+            sudo kubectl --kubeconfig /etc/kubernetes/admin.conf apply -f
+            /tmp/examplesolution_v1alpha1_versionserver_cr.yaml
```
This CR must be put in the right namespace `example-environment`.

Ebaneck replied:
https://github.com/scality/metalk8s/pull/2321/files#diff-fa1f5352a12d0b35290d51280da84271R5

alexandre-allard replied:
Oh! Yes! :p

alexandre-allard commented on eve/main.yml:
```diff
@@ -1466,3 +1469,277 @@ stages:
+          workdir: build/eve/workers/openstack-multiple-nodes/terraform/
+          haltOnFailure: true
+      - SetPropertyFromCommand:
+        name: Set Cluster IP as Property for example-versionserver
```
There is an issue with indentation here.

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,277 @@ stages:
+          workdir: build/eve/workers/openstack-multiple-nodes/terraform/
+          haltOnFailure: true
+      - ShellCommand:
+          name: Verify exampleversions Deployment
```
Looks very similar to the `ShellCommand` above, maybe we could factorize this ?

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,277 @@ stages:
+                exit 1
+              fi
+              sleep 3
+            done'
```
You're not executing the above command here, only executing the result of it.
Should be something like:
```suggestion
            bash -c '
            ATTEMPTS=0
            declare -a GET_REPLICAS=(
                sudo kubectl $KUBECONF get deployments example-versionserver
                -n example-environment -o custom-columns=:status.availableReplicas
            )
            REPLICAS=$("${GET_REPLICAS[@]}")
            until [ $REPLICAS -eq 3 ]; do
              ((ATTEMPTS++))
              if [ $ATTEMPTS -eq 10 ]; then
                echo "Could not obtain Replicas after ${ATTEMPTS} attempts" >&2
                exit 1
              fi
              sleep 3
              REPLICAS=$("${GET_REPLICAS[@]}")
            done'
```
Or even (more compact way):
```suggestion
            bash -c '
            declare -a GET_REPLICAS=(
                sudo kubectl $KUBECONF get deployments example-versionserver
                -n example-environment -o custom-columns=:status.availableReplicas
            )
            for ((attempts=0, replicas=$("${GET_REPLICAS[@]}"); replicas != 3; replicas=$("${GET_REPLICAS[@]}"), ++attempts)); do
              if [ $attempts -eq 10 ]; then
                echo "Could not obtain Replicas after $attempts attempts" >&2
                exit 1
              fi
              sleep 3
            done'
```
Also, note that you can't protect `$KUBECONF` variable with double quotes since you did put the flag and its value in the same variable, otherwise you end up with a single argument and it does not work.
I think you should only put the path `/etc/kubernetes/admin.conf` in this variable and rename it to `$KUBECONFIG` to be consistent.

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,277 @@ stages:
+          command: |
+            ssh -F $SSH_CONFIG bootstrap --
+            bash -c '
+            IP=$(sudo kubectl "$KUBECONF" get service example-versionserver -n example-environment -o custom-columns=:spec.clusterIP) &&
```
Same as above, will not work.

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,277 @@ stages:
+      - ShellCommand:
+          name: Verify versionserver http endpoints
+          env:
+            HTTP: "http://"
```
Should be name `SCHEME` and I'm not sure that we need an environment variable for this, normally it should never be changed.

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,277 @@ stages:
+          env:
+            HTTP: "http://"
+            IP: "%(prop:versionserver_cluster_ip)s"
+            SUB_URL: ":8080/version"
```
Same here, not sure the variable is useful, and it should not be named this way, at least have a `PORT` variable and `URL_PATH`

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,277 @@ stages:
+          command: >
+            ssh -F ssh_config bootstrap
+            sudo kubectl ${KUBECONF} taint nodes --all node-role.kubernetes.io/infra:NoSchedule-
+            sudo kubectl ${KUBECONF} taint nodes --all node-role.kubernetes.io/bootstrap:NoSchedule-
```
It will not work this way, everything will be put in a single kubectl command.

Ebaneck replied:
yep it does not

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,300 @@ stages:
+          command: >
+            ssh -F ssh_config bootstrap -- "
+            sudo kubectl --kubeconfig ${KUBECONFIG} taint nodes --all ${BOOTSTRAP_TAINT} &&
+            sudo kubectl --kubeconfig ${KUBECONFiG} taint nodes --all ${INFRA_TAINT}
```
Wouw, this one was hard to find, you have a typo:
```suggestion
            sudo kubectl --kubeconfig ${KUBECONFIG} taint nodes --all ${INFRA_TAINT}
```

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,300 @@ stages:
+              -o custom-columns=:spec.clusterIP
+            )
+            IP=$("${GET_IP[@]}")
+            echo $IP'
```
Why not just execute the command instead of this:
```suggestion
            ssh -F $SSH_CONFIG bootstrap --
            sudo kubectl --kubeconfig "$KUBECONFIG" get service
            example-versionserver -n example-environment
            -o custom-columns=:spec.clusterIP
```

Ebaneck replied:
Done

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,295 @@ stages:
+              REPLICAS=$("${GET_REPLICAS[@]}")
+            done'
+          workdir: build/eve/workers/openstack-multiple-nodes/terraform/
+          haltOnFailure: true
```
Not tested, but something like that should be better (less code duplication):
```suggestion
      - ShellCommand: &check_k8s_resource_column_ssh
          name: Verify example-solution-operator Deployment
          env: &_env_check_k8s_resource_column_ssh
            KUBECONFIG: "/etc/kubernetes/admin.conf"
            NAMESPACE: "example-environment"
            RESOURCE_TYPE: "deployment"
            RESOURCE_NAME: "example-solution-operator"
            RESOURCE_COLUMN: ":status.availableReplicas"
            RESOURCE_VALUE: "1"
            RETRIES: "10"
            SLEEP: "3"
          command: |
            ssh -F ssh_config bootstrap --
            bash -c '
            ATTEMPTS=0
            declare -a GET_COLUMN_VALUE=(
                sudo kubectl --kubeconfig "$KUBECONFIG" get "$RESOURCE_TYPE"
                "$RESOURCE_NAME" -n "$NAMESPACE"
                -o custom-columns="$RESOURCE_COLUMN"
            )
            CURRENT_VALUE=$("${GET_COLUMN_VALUE[@]}")
            until [ "$CURRENT_VALUE" -eq "$RESOURCE_VALUE" ]; do
              (( ATTEMPTS++ ))
              if [ "$ATTEMPTS" -eq "$RETRIES" ]; then
                echo "$RESOURCE_COLUMN still not equal to $RESOURCE_VALUE after $ATTEMPTS attempts" >&2
                exit 1
              fi
              sleep "$SLEEP"
              CURRENT_VALUE=$("${GET_COLUMN_VALUE[@]}")
            done'
          workdir: build/eve/workers/openstack-multiple-nodes/terraform/
          haltOnFailure: true
      - ShellCommand:
          <<: *check_k8s_resource_column_ssh
          name: Verify example-versionserver Deployment
          env:
            <<: *_env_check_k8s_resource_column_ssh
            RESOURCE_NAME: "example-versionserver"
            RESOURCE_VALUE: "3"
```

alexandre-allard replied:
Anyway, I'm not even sure the first approach or this one can work, using simple quotes will prevent environment variables from being expanded in the string passed to `bash -c`... and if you use double quotes, some variables will be expanded before the script got executed.
Maybe it would be better to put this in a dedicated `.sh`.

Ebaneck replied:
Hmm, i would prefer doing it without a dedicated shell script

alexandre-allard replied:
You can, you just need to use double quotes around the whole bash script and escape things that must not be expanded the first time.

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,269 @@ stages:
+            )
+            status_code=$("${GET_STATUS_CODE[@]}")
+            until [ "$status_code" -eq 200 ]; do
+              $status_code
```
The script will fail here for sure:
```
200: command not found
```

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,269 @@ stages:
+              $status_code
+              ((ATTEMPTS++))
+              if [ "$ATTEMPTS" -eq 10 ]; then
+                echo "Failed to obtain endpoint after ${ATTEMPTS} attempts" >&2
```
```suggestion
                echo "Failed to reach endpoint after ${ATTEMPTS} attempts" >&2
```

alexandre-allard commented on examples/metalk8s-solution-example/get_resource.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Not sure this is the best place to put this script as it's more related to CI than examples.
Should be put under `eve/`

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,269 @@ stages:
+            HOST_CR_PATH: "examples/metalk8s-solution-example/example-operator/deploy/crds/examplesolution_v1alpha1_versionserver_cr.yaml"
+            GET_RESOURCE: "examples/metalk8s-solution-example/get_resource.sh"
+          command: >
+            scp -F $SSH_CONFIG "${HOST_CR_PATH}" bootstrap:/tmp &&
```
You can copy multiple files using only 1 scp command (like cp):
```
scp -F "$SSH_CONFIG" "$HOST_CR_PATH" "$GET_RESOURCE" bootstrap:/tmp
```

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,269 @@ stages:
+            SLEEP: "3"
+          command: |
+            ssh -F ssh_config bootstrap
+            sudo bash /tmp/get_resource.sh
```
Here you're executing 2 commands, it must either be:
```suggestion
          command: |
            ssh -F ssh_config bootstrap \
            sudo bash /tmp/get_resource.sh
```
Or:
```suggestion
          command: >
            ssh -F ssh_config bootstrap
            sudo bash /tmp/get_resource.sh
```

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,269 @@ stages:
+            SSH_CONFIG: >-
+              eve/workers/openstack-multiple-nodes/terraform/ssh_config
+            HOST_CR_PATH: "examples/metalk8s-solution-example/example-operator/deploy/crds/examplesolution_v1alpha1_versionserver_cr.yaml"
+            GET_RESOURCE: "examples/metalk8s-solution-example/get_resource.sh"
```
I don't see the point of using environment variables for these.

alexandre-allard commented on eve/main.yml:
```diff
@@ -1553,3 +1556,275 @@ stages:
+            RESOURCE_VALUE=1
+            RETRIES=10
+            SLEEP=3
+            /tmp/get_resource.sh
```
Will not work, this file is not executable.

gdemonet reviewed (dismissed):
Why are we doing this as Eve steps using bash scripts, instead of extending the Python test suite?

TeddyAndrieux commented on eve/main.yml:
```diff
@@ -458,6 +459,7 @@ stages:
           stage_names:
             - single-node-install-rhel
             - multiple-nodes-centos
+            - multiple-nodes-example-solution-centos #remove me
```
Do we really need a new environment to test solution ?? ( For me no) Couldn't be part of the multi node we already have ?

alexandre-allard replied:
No, because we will do it in postmerge and we only want to test solution deployment, not necessarily the whole expansion.

TeddyAndrieux replied:
Hmm yes true, then you also need to add like for all other step the right env variable (and add these env variable on `add_final_stauts` like for all other steps)
```
_metalk8s_internal_info:
      junit_info: &_install_multi-node_solution_junit_info
....
```
To generate the right junit result file.

gdemonet replied:
I don't think it should be a "multi nodes" cluster. IMO, we want:

- one Bootstrap, to install Metal, and then import + run the Solution 
- one extra VM to run Cypress tests

This should also be the case for "single node" tests, so yeah, maybe merge them.

TeddyAndrieux replied:
Then you need to remove taint on the bootstrap node, but it's doable yes (or add toleration for example solution)

Ebaneck replied:
Removing taints on the Bootstrap node was handled here: https://github.com/scality/metalk8s/pull/2321/files#diff-73e74177595a435f88a0701e605cf08eR1686-R1698

TeddyAndrieux replied:
Hmmm then why you need multi node ?
Anyway I agree with @gdemonet comment https://github.com/scality/metalk8s/pull/2321#pullrequestreview-383610719 this test should be directly in pytest

TeddyAndrieux commented on eve/get_resource.sh:
```diff
@@ -0,0 +1,17 @@
+    if [ "${ATTEMPTS}" -eq "${RETRIES}" ]; then
+    echo "${RESOURCE_COLUMN} still not equal to ${RESOURCE_VALUE} after ${ATTEMPTS} attempts" >&2
+    exit 1
+    fi
```
```suggestion
    if [ "${ATTEMPTS}" -eq "${RETRIES}" ]; then
        echo "${RESOURCE_COLUMN} still not equal to ${RESOURCE_VALUE} after ${ATTEMPTS} attempts" >&2
        exit 1
    fi
```

Ebaneck committed:
tests: Add tests for Solutions CLI

This commit adds tests to validate the Solutions CLI

Formally, we added eve steps to automate the deployment and testing of
Solutions CLI.

In an attempt to expand our test-suite, we have rewritten the test-suite
in Python bdd

gdemonet commented on tests/post/features/example_solution.feature:
```diff
@@ -0,0 +1,22 @@
+        And pods with label 'app.kubernetes.io/name=example-versionserver' are 'Ready'
+        And we have 3 running pod labeled 'app.kubernetes.io/name=example-versionserver' in namespace 'example-environment'
+        When we check the http endpoint '/version' on port '8080' with IP 'spec.cluster_ip' for 'example-versionserver' in namespace 'example-environment'
+        Then service returns '200' status code
```
Those are not scenarii, merely static checks that could be in a `Given` or a `Then`.

You want to test the effect of running the CLI `solutions.sh`, build one scenario per command:
- Given no Solution is imported, And the config file is absent, When I run the `solutions.sh import ...` command, Then ...
- Given Solution toto is imported, And it's not marked as active, When I run `solutions.sh activate ...`, Then ...
- etc.

Do the same for creation of environment, adding a solution instance to an environment, and optionally (but really, that's not the point of those tests) check that the Operator works by creating a CR and checking that it deploys a component as expected.

Later iterations will also validate the upgrade of a Solution, but for now, let's focus on those simple actions. Also make sure to clean-up the Solution deployment after the test run (use a `session`-scoped fixture, add the teardown commands to it when you managed to setup something, like importing a solution, and make sure it runs on test completion, be it successful or not).

Ebaneck replied:
ok

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+    assert solution_environment_obj is None
+
+
+@then(parsers.parse("we have no activated Solution '{name}'"))
+def no_activated_solution(host, request, name):
+    with host.sudo():
+        output = host.check_output(
+            'salt-call --out=json pillar.get "{}"'.format(SOLUTION_PILLAR_KEY)
+        )
+    data = json.loads(output)['local']
+
+    assert name not in data
```
This function is a mostly a duplicate of `is_activated_solution()`, you should create an helper and use it in both functions.
And btw, you're not testing if the solution is activated, but if it's available.

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+
+
+@when("we remove Taints on Bootstrap node before deployment")
+def remove_bootstrapNode_taints(host):
```
```suggestion
def remove_bootstrap_node_taints(host):
```

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+@when(parsers.parse(
+    "I create a solution environment '{environment}' for Solution '{name}'"))
+def create_solution_environment(request, host, environment, name):
+    iso_root = request.config.getoption("--iso-root")
+    cmd = str(iso_root / "solutions.sh")
+    with host.sudo():
+        res = host.run(' '.join([
+            cmd,
+            'create-env --name {}'.format(environment),
+        ]))
+        assert res.rc == 0, res.stdout
```
Same here, code duplication, you could have an helper for this, with something like:
```
def run_solutions_command(request, host, args):
    iso_root = request.config.getoption("--iso-root")
    cmd = [str(iso_root / "solutions.sh")] + args
    with host.sudo():
        res = host.run(' '.join(cmd))
        assert res.rc == 0, res.stdout


@when(parsers.parse("I create a solution environment '{environment}'"))
def create_solution_environment(request, host, environment):
    run_solutions_command(request, host, ["create-env", "--name", environment])
```
And the `{name}` parameter is never used.

alexandre-allard commented on tests/post/features/example_solution.feature:
```diff
@@ -0,0 +1,31 @@
+        And we have no Solution environment 'example-environment'
+        And we have no activated Solution 'example-solution'
+        And we have no Solution Configuration file present
+        And we have no Solution deployment 'example-solution-operator' in environment 'example-environment'
+        And we have no Solution deployment 'example-solution-ui' in environment 'example-environment'
```
These 2 will always be true since you're testing `And we have no Solution environment 'example-environment'` just before.

alexandre-allard commented on tests/post/features/example_solution.feature:
```diff
@@ -0,0 +1,31 @@
+        And the Solution Configuration file is absent
+        When we import a Solution archive '/var/tmp/example-solution.iso'
+        Then Solution archive 'example-solution' is imported correctly
+        When I activate Solution 'example-solution' with version '0.1.0-dev'
+        Then Solution 'example-solution' is activated
+        When I create a solution environment 'example-environment' for Solution 'example-solution'
```
```suggestion
        When we activate Solution 'example-solution' version '0.1.0-dev'
        Then Solution 'example-solution' version '0.1.0-dev' is activated
        When we create a solution environment 'example-environment' for Solution 'example-solution'
```

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+    # Remove Taints on the Bootstrap node before deploying
+    # example-solution-operator else, the operator pods will not startup
+
+    with host.sudo():
+        res = host.run(' '.join([
+            'kubectl --kubeconfig=/etc/kubernetes/admin.conf ',
+            'taint nodes --all {}'.format(BOOTSTRAP_TAINT),
+        ]))
+        assert res.rc == 0, res.stdout
+
+    with host.sudo():
+        res = host.run(' '.join([
+            'kubectl --kubeconfig=/etc/kubernetes/admin.conf ',
+            'taint nodes --all {}'.format(INFRA_TAINT),
+        ]))
+        assert res.rc == 0, res.stdout
```
Why using 2 times `host.sudo()` and why are you using lists if you're not really splitting arguments ?
```
        res = host.run(' '.join([
            'kubectl' '--kubeconfig' '/etc/kubernetes/admin.conf',
            'taint' 'nodes' '--all', INFRA_TAINT,
        ]))
```
Or
```
        res = host.run((
            'kubectl --kubeconfig=/etc/kubernetes/admin.conf '
            'taint nodes --all {}'
        ).format(INFRA_TAINT))
```

Ebaneck replied:
Lets shoot for the split argument version

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+def no_solution_config(host, name):
+    with host.sudo():
+        if host.file(SOLUTION_CONFIGURATION_FILE).exists:
+            assert False, 'Solution Configuration file exist at {}'.format(
```
```suggestion
            assert False, 'Solution Configuration file exists at {}'.format(
```

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+    # 3. Solutions ConfigMap has new Solution imported
+
+    with host.sudo():
+        if not host.file(SOLUTION_CONFIGURATION_FILE).exists:
+            assert False, 'No Solution Configuration file exist at {}'.format(
+                        SOLUTION_CONFIGURATION_FILE
+                    )
```
Why not directly asserting ? (and weird indentation)

```suggestion
    assert host.file(SOLUTION_CONFIGURATION_FILE).exists, \
        'No Solution Configuration file exists at {}'.format(
            SOLUTION_CONFIGURATION_FILE
        )
```

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+                        SOLUTION_CONFIGURATION_FILE
+                    )
+    with host.sudo():
+        file_contents = host.file(SOLUTION_CONFIGURATION_FILE).content_string
+        if name not in file_contents:
```
This file is valid YAML, you could easily parse it to ensure the string is at the right place. :)

Ebaneck replied:
Would add a Todo here for some other time!

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+@given(parsers.parse("no Solution environment '{environment}' is available"))
+def is_absent_solution_environment(host, environment, k8s_client):
+    solution_environment = get_environment(k8s_client, environment)
+    if solution_environment is not None:
+        pytest.fail("Found Solution environment {}".format(environment))
+    return environment
```
Could be an assertion, and what's the return used for ?

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+    with host.sudo():
+        if host.file(SOLUTION_CONFIGURATION_FILE).exists:
+            assert False, 'Solution Config file exist at path {}'.format(
+                        SOLUTION_CONFIGURATION_FILE
+                    )
```
Really weird indentation here.

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+        file_contents = host.file(SOLUTION_CONFIGURATION_FILE).content_string
+        if name not in file_contents:
+            assert False, 'No Solution ISO found in Solution Config {}'.format(
+                        SOLUTION_CONFIGURATION_FILE, name
+                    )
```
indentation

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,398 @@
+    with host.sudo():
+        if host.file(SOLUTION_CONFIGURATION_FILE).exists:
+            assert False, 'Solution Configuration file exist at {}'.format(
+                        SOLUTION_CONFIGURATION_FILE
+                    )
```
identation

alexandre-allard commented on tests/post/features/example_solution.feature:
```diff
@@ -0,0 +1,31 @@
+        Then solution environment 'example-environment' is available
+        When we remove Taints on Bootstrap node before deployment
+        And we deploy Solution 'example-solution' in environment 'example-environment' with version '0.1.0-dev'
+        Then the 'pods' list should not be empty in the 'example-environment' namespace
+        When we check the available replicas 'status.available_replicas' for 'example-solution-operator' deployment in environment 'example-environment'
+        Then we have '1' available 'status.available_replicas' for Solution 'example-solution-operator' Deployment in environment 'example-environment'
+        When we check the available replicas 'status.available_replicas' for 'example-solution-ui' deployment in environment 'example-environment'
+        Then we have '1' available 'status.available_replicas' for Solution 'example-solution-ui' Deployment in environment 'example-environment'
```
```suggestion
        Then the 'pods' list should not be empty in the 'example-environment' namespace
        And we have '1' available 'status.available_replicas' for Solution 'example-solution-operator' Deployment in environment 'example-environment'
        And we have '1' available 'status.available_replicas' for Solution 'example-solution-ui' Deployment in environment 'example-environment'
```

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,331 @@
+    return response.items
+
+
+def get_environment(k8s_client, name):
+    try:
+        response = k8s_client.read_namespace(name)
+    except (ApiException) as exc:
+        if isinstance(exc, ApiException) and exc.status == 404:
+            return None
+        raise
+    return response
```
An environment should not be matched on the namespace name but on a specific label `metadata:labels:solutions.metalk8s.scality.com/environment` with the environment name as the value.
So you must use `list_namespace()` with a label selector, should be something like that:
```
ENVIRONMENT_LABEL = "solutions.metalk8s.scality.com/environment"
response = k8s_client.list_namespace(label_selector='{}={}'.format(ENVIRONMENT_LABEL, name))
```

Ebaneck replied:
Ok, i will fix this!

alexandre-allard commented on tests/post/features/example_solution.feature:
```diff
@@ -0,0 +1,27 @@
+        And no Solution environment 'example-environment' is available
+        And the Solution Configuration file is absent
+        When we import a Solution archive '/var/tmp/example-solution.iso'
+        Then Solution archive 'example-solution' is imported correctly
+        When we activate Solution 'example-solution' version '0.1.0-dev'
+        Then Solution 'example-solution' is available
```
```suggestion
        Then Solution archive 'example-solution' is imported correctly
        And Solution 'example-solution' version '0.1.0-dev' is available
        When we activate Solution 'example-solution' version '0.1.0-dev'
        Then Solution 'example-solution' version '0.1.0-dev' is activated
        And CRD 'versionservers.example-solution.metalk8s.scality.com' exists in Kubernetes API
        And CRD 'clockservers.example-solution.metalk8s.scality.com' exists in Kubernetes API
```

alexandre-allard replied:
The Solution is made available at import, not at activation.

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,331 @@
+def is_absent_solution(host, name, k8s_client):
+    with host.sudo():
+        assert name not in host.mount_point.get_mountpoints(), \
+            'No Solution Configuration file exists at {}'.format(name)
```
The error message is not related to what's asserted.

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,331 @@
+
+
+@when("we remove Taints on Bootstrap node before deployment")
+def remove_bootstrap_node_taints(request, host):
+    # Remove Taints on the Bootstrap node before deploying
+    # example-solution-operator else, the operator pods will not startup
+
+    with host.sudo():
+        res = host.run((
+            'kubectl --kubeconfig=/etc/kubernetes/admin.conf '
+            'taint nodes --all {}'
+        ).format(BOOTSTRAP_TAINT))
+        assert res.rc == 0, res.stdout
+
+    with host.sudo():
+        res = host.run((
+            'kubectl --kubeconfig=/etc/kubernetes/admin.conf '
+            'taint nodes --all {}'
+        ).format(INFRA_TAINT))
+        assert res.rc == 0, res.stdout
```
Should either be `remove_nodes_taints()` or better: you should only remove taints on bootstrap node here (not using the `--all flag`), otherwise, if someone runs this scenario on a multi nodes cluster, it'll have unintended behavior.

alexandre-allard commented on tests/post/features/example_solution.feature:
```diff
@@ -0,0 +1,27 @@
+@local @solution
+Feature: Example solution
```
Feature is not "Example solution", but "Solutions CLI" or just "Solutions", file should be also renamed accordingly.

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,331 @@
+    assert name not in get_available_solution(host)
+
+
+@then(parsers.parse("we have no Solution Configuration file present"))
+def no_solution_config(host, name):
```
No need for parsers.parse() if there is nothing to parse.
And from where does the `name` parameter comes from ?

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,331 @@
+    return response
+
+
+def list_environments(k8s_client):
+    try:
+        response = k8s_client.list_namespace()
+    except Exception as exc:
+        pytest.fail("Unable to read namespaces: {!s}".format(exc))
+    return response.items
```
This function is not used anywhere.

alexandre-allard commented on tests/utils.py:
```diff
@@ -66,3 +72,111 @@ def get_node_name(nodename, ssh_config=None):
     return nodename
+
+
+# Source: https://www.peterbe.com/plog/best-practice-with-retries-with-requests
+def requests_retry_session(
+    retries=3,
+    backoff_factor=0.3,
+    status_forcelist=(500, 503),
+    method_whitelist=frozenset(['GET', 'POST']),
+    session=None
+):
+    """Configure a `requests.session` for retry on error.
+
+    By default, this helper performs 3 retries with an exponential sleep
+    interval between each request and only retries internal server errors(500)
+    & service unavailable errors(503)
+
+    Arguments:
+        retries:          The number of retries to perform before giving up
+        backoff_factor:   The sleep interval between requests computed as
+                          {backoff factor} * (2 ^ ({number retries} - 1))
+        status_forcelist: HTTP status codes that we should force a retry on
+        method_whitelist: uppercased HTTP methods that we should retry
+        session:          Used to create a session
+
+    Returns:
+        A `requests.Session` object configured for retry.
+    """
+    session = session or requests.Session()
+    retry = Retry(
+        total=retries,
+        read=retries,
+        connect=retries,
+        backoff_factor=backoff_factor,
+        status_forcelist=status_forcelist,
+    )
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount("http://", adapter)
+    session.mount("https://", adapter)
+    return session
```
Not used

Ebaneck replied:
Yes, will remove this one!
It was a cherry-pick from an old commit

alexandre-allard commented on tests/conftest.py:
```diff
@@ -150,6 +150,14 @@ def ssh_config(request):
     return request.config.getoption('--ssh-config')
 
 
+@pytest.fixture(scope="function")
+def request_retry_session(request):
+    # Callers can inject arguments using `pytest.mark.parametrize`
+    params = getattr(request, 'param', {})
+
+    return utils.requests_retry_session(**params)
+
+
```
not used

alexandre-allard commented on tests/utils.py:
```diff
@@ -66,3 +72,111 @@ def get_node_name(nodename, ssh_config=None):
+
+    if kwargs.get('container'):
+        kube_args.extend(['-c', kwargs.get('container')])
+    if kwargs.get('namespace'):
+        kube_args.extend(['-n', kwargs.get('namespace')])
+
+    kubectl_cmd_tplt = 'kubectl exec {} {} -- {}'
+
+    with host.sudo():
+        output = host.run(
+            kubectl_cmd_tplt.format(
+                pod,
+                ' '.join(kube_args),
+                ' '.join(command)
+            )
+        )
+        return output
+
+
+def run_salt_command(host, command, ssh_config):
+    """Run a command inside the salt-master container."""
+
+    pod = 'salt-master-{}'.format(
+        get_node_name('bootstrap', ssh_config)
+    )
+
+    output = kubectl_exec(
+        host,
+        command,
+        pod,
+        container='salt-master',
+        namespace='kube-system'
+    )
+
+    assert output.exit_status == 0, \
+        'command {} failed with: \nout: {}\nerr:'.format(
+            command,
+            output.stdout,
+            output.stderr
+        )
```
These two functions are not used either.

alexandre-allard commented on tests/utils.py:
```diff
@@ -66,3 +72,111 @@ def get_node_name(nodename, ssh_config=None):
+    return functools.reduce(operator.getitem, path.split(delimiter), data)
+
+
+def set_dict_element(data, path, value, delimiter='.'):
+    """
+    Traverse a nested dict using a delimiter on a target string
+    replaces the value of a key within a dictionary and returns the new dict
+    """
+    path, _, key = path.rpartition(delimiter)
+    (get_dict_element(data, path) if path else data)[key] = value
+    return data
```
Not used

Ebaneck replied:
It is a cherry-pick from a commit that was only merged in 2.5
Atleast we use the `get_dict_element` method, so!

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,331 @@
+    try:
+        response = k8s_client.read_namespace(name)
+    except (ApiException) as exc:
+        if isinstance(exc, ApiException) and exc.status == 404:
```
Not sure the `isinstance(exc, ApiException)` is needed.

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,331 @@
+        response = k8s_client.read_namespaced_config_map(
+            name, namespace
+        ).to_dict()
+    except Exception as exc:
+        pytest.fail(
+            "Unable to read {} ConfigMap  in namespace {} with error: {!s}"
+            .format(name, namespace, exc)
+        )
```
Why not doing the same as for `get_environment()`, returning an empty dict for a `404` status code and letting the caller decide what to do when the `ConfigMap` cannot be found.

alexandre-allard commented on tests/post/steps/test_example_solution.py:
```diff
@@ -0,0 +1,331 @@
+    run_solutions_command(request, host, ["unimport", "--archive", path])
+
+
+@when(parsers.parse("we delete the Solution Configuration file"))
```
no need for `parsers.parse()`

alexandre-allard commented on tests/post/steps/test_solutions.py:
```diff
@@ -0,0 +1,383 @@
+
+@then(parsers.parse("solution environment '{name}' is available"))
+def read_solution_environment(k8s_client, name):
+    solution_environment_obj = get_environment(k8s_client, name)
+    assert any(name == env.metadata.name for env in solution_environment_obj)
```
Environment name != namespace name, anyway it should be:
```suggestion
    assert get_environment(k8s_client, name) is not None
```

alexandre-allard commented on tests/post/steps/test_solutions.py:
```diff
@@ -0,0 +1,383 @@
+
+@then(parsers.parse("we have no Solution environment '{name}'"))
+def no_solution_environment(k8s_client, name):
+    solution_environment_obj = get_environment(k8s_client, name)
+    assert name not in solution_environment_obj
```
```suggestion
    assert get_environment(k8s_client, name) is None
```
And you should write explicit assertion error message.

alexandre-allard commented on tests/post/steps/test_solutions.py:
```diff
@@ -0,0 +1,383 @@
+        available_solutions = get_available_solution(host)
+        for k, v in available_solutions.items():
+            if k == name:
+                for items in v:
+                    assert bool(items.get('active', None))
+                    assert items.get('version', None) == solution_version
```
You should not do it that way, here it'll be okay because we ensure there is only one Solution deployed at a time, but you should not rely on it:
```suggestion
                assert any([item.get('active') and item.get('version') == solution_version for item in v]) is True, \
                    "Solution {} in version {} is not active".format(name, solution_version)
```

alexandre-allard-scality commented on tests/post/steps/test_solutions.py:
```diff
@@ -0,0 +1,383 @@
+            'kubectl --kubeconfig=/etc/kubernetes/admin.conf '
+            'get crd {}'
+        ).format(name))
+        assert res.rc == 0, res.stdout
```
Please write assertion error messages.

Ebaneck replied:
`Stdout` is sufficient no??

alexandre-allard commented on tests/post/features/solutions.feature:
```diff
@@ -0,0 +1,30 @@
+        And we have '1' available 'status.available_replicas' for Solution 'example-solution-operator' Deployment in environment 'example-environment'
+        And we have '1' available 'status.available_replicas' for Solution 'example-solution-ui' Deployment in environment 'example-environment'
+        When we deactivate Solution 'example-solution'
+        And we delete Solution 'example-solution' in environment 'example-environment'
+        And we delete Solution environment 'example-environment'
```
If we delete a Solution in an Environment and then delete the Environment right after, without checking if the Solution was successfully removed, we can't ensure the first command has worked as expected.

alexandre-allard-scality commented on tests/post/features/solutions.feature:
```diff
@@ -0,0 +1,30 @@
+        When we remove Taints on node 'bootstrap' before deployment
+        And we deploy Solution 'example-solution' in environment 'example-environment' with version '0.1.0-dev'
+        Then the 'pods' list should not be empty in the 'example-environment' namespace
+        And we have '1' available 'status.available_replicas' for Solution 'example-solution-operator' Deployment in environment 'example-environment'
+        And we have '1' available 'status.available_replicas' for Solution 'example-solution-ui' Deployment in environment 'example-environment'
```
You could have used what's already done for the rest of the tests (defined in conftest.py).
```
we have 1 running pod labeled 'name=example-solution-operator' in namespace 'example-environment'
we have 1 running pod labeled 'app=example-solution-ui' in namespace 'example-environment'
```

alexandre-allard commented on tests/post/features/solutions.feature:
```diff
@@ -0,0 +1,30 @@
+        Then solution environment 'example-environment' is available
+        When we remove Taints on node 'bootstrap' before deployment
+        And we deploy Solution 'example-solution' in environment 'example-environment' with version '0.1.0-dev'
+        Then the 'pods' list should not be empty in the 'example-environment' namespace
+        And we have '1' available 'status.available_replicas' for Solution 'example-solution-operator' Deployment in environment 'example-environment'
+        And we have '1' available 'status.available_replicas' for Solution 'example-solution-ui' Deployment in environment 'example-environment'
```
You could have used what's already done for the rest of the tests (defined in conftest.py).

```suggestion
        Then we have 1 running pod labeled 'name=example-solution-operator' in namespace 'example-environment'
        And we have 1 running pod labeled 'app=example-solution-ui' in namespace 'example-environment'
```

Ebaneck replied:
Done

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI` with contents from `feature/1856-add-Integration-of-Solution-tests-in-CI`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI origin/development/2.5
 $ git merge origin/feature/1856-add-Integration-of-Solution-tests-in-CI
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI
```

alexandre-allard commented on tests/post/steps/test_solutions.py:
```diff
@@ -0,0 +1,350 @@
+    for k, v in available_solution.items():
+        if k == name:
+            for item in v:
+                assert item.get('version', None) == solution_version, \
+                    'Expected solution version {}: got {}'.format(
+                        solution_version, item.get('version', None)
+                    )
```
Not the good way of doing this, you should check each item to see if there is a matching version.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1856-add-Integration-of-Solution-tests-in-CI` into
`development/2.4`
* [w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI) will be merged into `development/2.5` 
* [w/2.6/feature/1856-add-Integration-of-Solution-tests-in-CI](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/1856-add-Integration-of-Solution-tests-in-CI) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

Ebaneck commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/b941ef939977b95636516086740cd8c2ee6705eb) did not succeed in branch w/2.5/feature/1856-add-Integration-of-Solution-tests-in-CI.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ebaneck.
